File: src/main/scala/scalaexpr/Expressions.scala
package scalaexpr

sealed abstract class Expression

case class Number(value: Int) extends Expression
case class BinOp(op: Op, left: Expression, right: Expression) extends Expression

sealed abstract class Op
case object Plus extends Op
case object Minus extends Op
case object Times extends Op
case object Div extends Op

object Expressions {
  
  /** Evaluation */
  def eval(e: Expression): Int = e match {
    case Number(n) => n
    case BinOp(op, e1, e2) =>
      val v1 = eval(e1)
      val v2 = eval(e2)
      op match {
        case Plus => v1 + v2
        case Minus => v1 - v2
        case Times => v1 * v2
        case Div => v1 / v2
      }
  }
 
  /** Formating */
  def prec(e: Expression): Int = e match {
    case Number(_) => 0
    case BinOp(Times | Div, _, _) => 1
    case BinOp(Plus | Minus, _, _) => 2
  }
  
  def format(op: Op): String = op match {
    case Plus => " + "
    case Minus => " - "
    case Times => " * "
    case Div => " / "
  }
  
  def format(e: Expression): String = e match {
    case Number(n) => n.toString()
    case BinOp(op, e1, e2) =>
      def paren(left: Boolean, ep: Expression, e: Expression): String = {
        if (prec(ep) < prec(e) || !left && prec(ep) == prec(e)) {
          "(" + format(e) + ")" 
        } else {
          format(e)
        }
      }
      paren(true, e, e1) + format(op) + paren(false, e, e2)
  }

  /** Simplification */
  def simplifyTop(e: Expression): Expression = e match {
    case BinOp(Plus, e1, Number(0)) => e1
    case BinOp(Times, _, e2 @ Number(0)) => e2
    case BinOp(Times, e1, Number(1)) => e1
    case BinOp(Times, e1, e2) if e1 == e2 => BinOp(Times, Number(2), e1)
    case _ => e
  }
   
}

File: src/main/java/javaexpr/visitor/Expressions.java
package javaexpr.visitor;

/** The Visitor class representing a generic operation on expressions */
abstract class Visitor<T> {
	abstract T visit(Number number);
	
	T visit(BinOp binop) {
		return binop.op().dispatch(this, binop.left(), binop.right());
	}
	
	abstract T visit(Plus op, Expression e1, Expression e2);
	abstract T visit(Minus op, Expression e1, Expression e2);
	abstract T visit(Times op, Expression e1, Expression e2);
	abstract T visit(Div op, Expression e1, Expression e2);
}

/** The abstract expression type */
abstract class Expression {
  abstract <T> T dispatch(Visitor<T> visitor);
}	

/** Number expressions */
class Number extends Expression {
		
  private final int value;
	    
  Number(int value) {
	  this.value = value;
  }
 
  int value() {
	  return value;
  }

  <T> T dispatch(Visitor<T> visitor) {
  	return visitor.visit(this);
  }
  
}

/** The abstract operator type */
abstract class Op {
  abstract <T> T dispatch(Visitor<T> visitor, Expression e1, Expression e2);
}

/** The addition operator */
class Plus extends Op {
	<T> T dispatch(Visitor<T> visitor, Expression e1, Expression e2) {
		return visitor.visit(this, e1, e2);
	}
}

/** The subtraction operator */
class Minus extends Op {
	<T> T dispatch(Visitor<T> visitor, Expression e1, Expression e2) {
		return visitor.visit(this, e1, e2);
	}
}

/** The multiplication operator */
class Times extends Op {
	<T> T dispatch(Visitor<T> visitor, Expression e1, Expression e2) {
		return visitor.visit(this, e1, e2);
	}
}

/** The division operator */
class Div extends Op {
	<T> T dispatch(Visitor<T> visitor, Expression e1, Expression e2) {
		return visitor.visit(this, e1, e2);
	}
}

/** Binary operator expressions */
class BinOp extends Expression {

	private final Expression left;
	private final Op op;
	private final Expression right;

	BinOp(Expression left, Op op, Expression right) {
    this.left = left;
    this.op = op;
    this.right = right;
	}

	Expression left() {
	  return left;
	}

	Op op() {
	  return op;
	}

	Expression right() {
	  return right;
	}

	<T> T dispatch(Visitor<T> visitor) {
  	return visitor.visit(this);
  }
}

/** A visitor for evaluating expressions */
class Evaluator extends Visitor<Integer> {
	
  Integer visit(Number e) { return e.value(); }

  Integer visit(Plus op, Expression e1, Expression e2) {
	  return e1.dispatch(this) + e2.dispatch(this);
  }
  
  Integer visit(Minus op, Expression e1, Expression e2) {
  	return e1.dispatch(this) - e2.dispatch(this);
  }
  
  Integer visit(Times op, Expression e1, Expression e2) {
  	return e1.dispatch(this) * e2.dispatch(this);
  }
  
  Integer visit(Div op, Expression e1, Expression e2) {
  	return e1.dispatch(this) / e2.dispatch(this);
  }
}

/** A visitor for formating expressions */
class Formatter extends Visitor<Void> {
	
	StringBuffer buffer = new StringBuffer();
	
  Void visit(Number e) { 
  	buffer.append(e.value());
  	return null;
  }

  Void visit(Plus op, Expression e1, Expression e2) {
	  e1.dispatch(this);
	  buffer.append(" + ");
	  e2.dispatch(this);
	  return null;
  }
  
  Void visit(Minus op, Expression e1, Expression e2) {
	  e1.dispatch(this);
	  buffer.append(" - ");
	  e2.dispatch(this);
	  return null;
  }
  
  Void visit(Times op, Expression e1, Expression e2) {
	  e1.dispatch(this);
	  buffer.append(" * ");
	  e2.dispatch(this);
	  return null;
  }
  
  Void visit(Div op, Expression e1, Expression e2) {
	  e1.dispatch(this);
	  buffer.append(" / ");
	  e2.dispatch(this);
	  return null;
  }
  
  public String toString() {
  	return buffer.toString();
  }
}

/** For testing ... */
public class Expressions {	  
  public static int eval(Expression e) {
  	return e.dispatch(new Evaluator());
  }
	
  public static String format(Expression e) {
  	Formatter f = new Formatter();
  	e.dispatch(f);
    return f.toString();
  }
  
	public static void main(String[] args) {
  	// Create an expression for 2 * 3 + 10 / 5.
	  Expression e =
	    new BinOp(new BinOp(new Number(2), new Times(), new Number(3)),
		            new Plus(),	  
			  	      new BinOp(new Number(10), new Div(), new Number(5)));

    // Print the expression, evaluate it, and print the result.
	  System.out.println(format(e) + " = " + eval(e));
  }
}
